### PR TITLE
Added EitherCollectors

### PR DESCRIPTION
**Pull request by peterunold**
Hi - perhaps you'd be interested in pulling this addition of Collectors to your Either project. Basically the EitherCollectors class adds a left and a right biased collectors for streams of Either<> objects:

```
List<Either<String,Integer>> listOfEithers = ...;

Either<List<String>,List<Integer>>  eitherList = listOfEithers.collect(EitherCollectors.toLeftBiasedList());
```
I've been using it in a couple of projects and found it useful for batch processing lists of requests. If one request cannot be processed the entire batch is rejected.  

I'd be happy to write tests for it.

### CODE REVIEW

**Comment by peterunold** [comment]
> This is a really nice addition! I'd be down for rolling it into the upstream repo and releasing it to Maven Central in a new version, if you want to add a JUnit test and PR it up. I'll even give Hacktoberfest credit for ya if you'd like.

Oh I didn't see your comment before I made the PR. I'll add some tests and make a new PR.

**Comment by peterunold** [comment]
@spencerwi Added Java tests for the collectors. I believe the Kotlin sequences don't use collectors?

File: src/main/java/com/spencerwi/either/EitherCollectors.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+    * Factory method for creating a left biased collector
+    */
+    public static <L,R> Collector<Either<L,R>, ?, Either<List<L>, List<R>>> toLeftBiasedList() {
+        return new EitherCollectors<>(true);
+    }
+
+    /**
+     * Factory method for creating a right biased collector
+     */
```
Comment (spencerwi) [inline src/main/java/com/spencerwi/either/EitherCollectors.java:36]: I think it'd be useful to move an explanation of what a "left-biased" or "right-biased" list is here.

In reading through the code, I had a hard time wrapping my head around it -- the most confusing piece here, I think, is that this Collector doesn't actually produce a `List`, but instead produces an `Either<List<L>, List<R>>`

Could you walk me through an example use-case for this collector?

File: src/test/java/com/spencerwi/either/EitherCollectorsTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisplayName("EitherCollectors with Java")
+public class EitherCollectorsTest {
+
+    @Test
+    public void checkBiasedCollectors(){
+        Either<String, Integer> either1 = Either.left("left");
+        Either<String, Integer> either2 = Either.right(42);
```
Comment (spencerwi) [inline src/test/java/com/spencerwi/either/EitherCollectorsTest.java:19]: I think this test would benefit quite a bit from having multiple of each "side" in the list. That would not only better exercise the collector code, but would also serve as a clearer example of what this collector does.

Maybe have a list that goes `Left, Right, Right, Left, Left`, and then assert that the results of each biased-collector are correct.

File: src/main/java/com/spencerwi/either/EitherCollectors.java (hunk 1 above)
Comment (peterunold) [reply src/main/java/com/spencerwi/either/EitherCollectors.java:37]: I've primarily been using the collectors when processing lists of requests or validating lists of expressions. Here I want all requests or expression to be right before proceeding or reject the entire list. 

Example:

```
Either<List<String>, List<ParsedExpression>> = expressionList.stream()
  .map(expr -> parseExpression(expr))
  .collect(EitherCollectors.toLeftBiased());
```
At this point, if one of the expressions was illegal(return a left-either) we want to reject the entire list and report back the error message for each failed expression.

**Comment by spencerwi** [comment]
Thanks for the example use-case, and for the changes! I'll update the README, bump the version, and release once I get a chance. Thanks again!

**Comment by peterunold** [comment]
You're very welcome. Your Either project has been a joy to use across multiple projects.

**Pull request merged**